Fix hydra integration and make tests succeed

dotta opened the pull request:
Mirrors https://github.com/scalacenter/bloop/pull/1184

This PR should fix the execution of tests with Hydra. It includes two important fixes:

1) Thread-safe analysis callback, which is required by Hydra since several compilation threads may be stepping on each others.
2) Updated the bloop-hydra-bridge version as it contains an important fix to
correctly handle invalidated classfiles.

For more details, please have a look at the commit messages.

Aspects that still need to be improved:

1) Shall we rename `IBloopAnalysisCallback` to something else?
2) A thread-safe Reporter is likely need for correct execution with Hydra. But switching the mutable data structures used in the different Reporter implementations into concurrent ones might be a first good approximation.

jvican commented on backend/src/main/scala/bloop/reporter/Reporter.scala:
```diff
@@ -30,6 +30,8 @@ abstract class Reporter(
     override val config: ReporterConfig,
+    // FIXME: _problems needs to be thread-safe. Shall we use a Java ConcurrentLinkedQueue or would you prefer something else?
+    //        Keeping the mutable buffer can cause Hydra to crash because of off-by-one exceptions thrown for instance by Buffer#clear
     val _problems: mutable.Buffer[ProblemPerPhase] = mutable.ArrayBuffer.empty
```
Hmmm, would it be possible to parameterize the reporter with a buffer interface of sorts so that we can switch the implementation to `ConcurrentLinkedQueue` when using Hydra?

dotta replied:
Sounds good. I'm going to have an abstract Buffer interface with two implementations: one using the current Scala `mutable.Buffer` and another one using a `ConcurrentLinkedQueue`.

jvican replied:
Yes, I think having a simple interface with only the methods we use should be enough, thanks Mirco 👍

dotta replied:
Yep, that's exactly what I have. The implementations will be value classes, so we should have no allocation penalty (not that I think it really matters here, but I thought you'd still like that :)).

jvican replied:
Great, another way we might want to implement this is by avoiding creating an interface for a buffer and just having some abstract methods to add/manage problems in the reporter interface that are implemented in a `SingleThreadedReporter` and a `ConcurrentReporter`. I almost find this solution better than creating an interface for buffers. What do you think?

dotta replied:
I've just pushed the changes that implement the solution we initially discussed. TBH, I'd actually suggest to make the concurrent implementation the default, as I don't believe users will experience any performance penalty, and it would keep the code simpler (as there is only one implementation and all `Reporter` implementations simply need to be thread-safe)

jvican replied:
> I'd actually suggest to make the concurrent implementation the default
  
I prefer to err on the side of caution and given that we don't have the time to run benchmarks to make sure this is the case that we keep these two implementations separate

dotta replied:
Sure. I'm totally ok with keeping things as they are. Shall we mark this conversation resolved?

jvican commented on backend/src/main/scala/bloop/reporter/Reporter.scala:
```diff
@@ -38,7 +40,10 @@ abstract class Reporter(
+  // FIXME: Adding a synchronized here is clearly not the proper fix, but it seems to help preventing random failures of the test. As
+  //        I'd like to see the test suite succeed on the CI, I've decided to include it in the current PR, even though we will of course
+  //        remove this before merging it.
+  override def reset(): Unit = synchronized {
```
I wonder why `reset` is being called concurrently while compiling a project. I would expect Hydra doesn't call it unless all threads are done compiling a project. I'm concerned the fact we need `synchronized` might be making some other parts in Bloop not work well with Hydra, what do you think @dotta?

dotta replied:
You are correct that `reset` is not called concurrently. Maybe the race condition I was seeing was an artifact of the tests, but unfortunately I'm unable to reproduce the exception (I believe the exception I was seeing was an `ArrayIndexOutOfBoundException` with index -1).

Regardless, since errors may be reported concurrently, we still need to make the underlying `_problems` data structure thread-safe. That's exactly what I'm doing and the `synchronized` will disappear.

dotta replied:
Jorge, the call to `reporter.reset` is done in [`Global`](https://github.com/scala/scala/blob/2.13.x/src/compiler/scala/tools/nsc/Global.scala#L1493). That's why the race condition with the `Reporter._problems` could surface when compilation is carried out with  Hydra.

jvican replied:
Right, so it is indeed being called concurrently when the compiler is preparing the compilation units. Any chance you can change this in Hydra?

dotta replied:
Why would this need to be changed in Hydra and not in the vanilla Scala compiler too? The approach I'm taking is to make the Reporters implementation thread-safe, as we discussed.

jvican replied:
Even if you make the implementation thread-safe, could we not have a race condition here? If one hydra compile run finishes early before other compile runs and calls `reset` then the problems diagnosed by the other runs could be disappeared, no?

dotta replied:
`reset` is not called by Hydra. And because the current implementation works when used with the vanilla Scala compiler, there should be no reason for it to not work with Hydra.

dotta replied:
To clarify, `Reporter.reset` is called in Global as mentioned earlier, but this call happens _before_ the sources are compiled. So, while it's true there is a chance of a race-condition, empirical evidence from our experience with Hydra is that this call to `Reporter.reset` is generally executed across all Hydra threads before a problem is reported.

jvican replied:
> So, while it's true there is a chance of a race-condition, empirical evidence from our experience with Hydra is that this call to Reporter.reset

Sure, so long as we're explicit about this and we acknowledge that this race condition could happen in the future or could be reported by users, I'm fie with it :smile:

dotta replied:
Sounds good. I've actually filed a ticket on our side as I think this is should be easy to fix (and I think we can have a fix included in the next Hydra release). I'm going to marke this discussion resolved as there is no further action to take here.

jvican commented on backend/src/main/scala/sbt/internal/inc/bloop/internal/BloopIncremental.scala:
```diff
@@ -51,7 +52,11 @@ object BloopIncremental {
     val internalSourceToClassNamesMap: File => Set[String] = (f: File) => previousRelations.classNames(f)
 
-    val builder = () => new BloopAnalysisCallback(mode, internalBinaryToSourceClassName, internalSourceToClassNamesMap, externalAPI, current, output, options, manager)
+    val builder: () => IBloopAnalysisCallback = {
```
:+1: Looks great!

dotta commented:
Thanks @jvican! Will work on this right away. By the way, I think you got my github handle wrong ;-)

dotta commented on backend/src/main/scala/sbt/internal/inc/bloop/internal/ThreadSafeAnalysisCallback.scala:
```diff
@@ -0,0 +1,398 @@
+ *
+ * IMPORTANT: All modifications made to BloopAnalysisCallback` must be replicated here.
+ */
+final class ThreadSafeAnalysisCallback(
```
A better name would be `ConcurrentAnalysisCallback`. I'm happy to make the change if you'd like so.

dotta replied:
Alright, I'll do it if there are other changes that are needed (which I'm not sure there will be). So, this shouldn't be blocking a merge if you are happy with the PR.

dotta commented:
@jvican I've rebased on master (as I spotted there was a commit addressing flaky tests) and essentially the last two commits contain the new code to review (addressing thread-safety for all Reporter implementations). Please, read the commit messages for details.

jvican commented on frontend/src/main/scala/bloop/reporter/BspProjectReporter.scala:
```diff
@@ -16,22 +16,40 @@ import scala.collection.mutable
 
   /** A thread-safe map with all the files under compilation. */
-  private val compilingFiles = mutable.HashMap.empty[File, Boolean]
+  private val compilingFiles = TrieMap.empty[File, Boolean]
```
Let's use `ConcurrentHashMap` instead of `TrieMap`

dotta replied:
I've used a `TrieMap` for coherence with the other fields defined in the same class (see [here](https://github.com/scalacenter/bloop/blob/master/frontend/src/main/scala/bloop/reporter/BspProjectReporter.scala#L37-L40)). But if you prefer a `ConcurrentHashMap` just for this field, I'm of course happy to make the change. Let me know!

jvican replied:
Yes, please, let's use `ConcurrentHashMap` here, this field is accessed quite often so I prefer to use a more efficient implementation than `TrieMap`

dotta replied:
You got it. Will mark this resolved.

dotta commented on frontend/src/main/scala/bloop/reporter/BspProjectReporter.scala:
```diff
@@ -16,22 +16,40 @@ import scala.collection.mutable
 
   /** A cycle count, initialized to 0 when it's a no-op. */
-  private var cycleCount: Int = 0
+  private val cycleCount: AtomicInt = AtomicInt(0)
```
Yes, otherwise there can be visibility problems due to the `Reporter` instance being accessed from different threads. Also, the commit message in https://github.com/scalacenter/bloop/pull/1190/commits/ae26519b437881291880c116740e5d649dcdd63b provides context on this change (in case you missed it)

dotta replied:
I'm going to mark this resolved.

jvican commented:
> It's worth mentioning that the `Reporter.nextID` while being now thread-safe,
may actually not return a valid ID when compilation is done with Hydra.
  
I think that this might cause some important problems, `nextID` generates a unique ID per problem and IIRC this id is relied on by clients to deduplicate problems, so it's quite critical the uniqueness property is maintained. What would it take to satisfy that?
  
Otherwise, this is looking neat! :+1: Thanks for the hard work

dotta commented:
> I think that this might cause some important problems, nextID generates a unique ID per problem and IIRC this id is relied on by clients to deduplicate problems, so it's quite critical the uniqueness property is maintained. What would it take to satisfy that?

Actually, I believe I was too negative. `nextID` exposes the same race-condition we discussed for `Reporter.reset` (which I mentioned we should be able to fix on the Hydra side). Other than that, the returned ID will be guaranteed to be unique since it's generated with an atomic increment. So, in short, I'll update the commit message to match what I said now.

dotta commented:
@jvican I believe I've addressed all comments. Here are the set of changes I've just pushed:

* Renamed `ThreadSafeAnalysisCallback` to `ConcurrentAnalysisCallback`.
* Updated [this commit message](https://github.com/scalacenter/bloop/pull/1190/commits/6e555f9817c1f83e9b44fd6c0d779b63f7bdda7c) to correct my incorrect description of concurrency problems related to `Reporter.nextID` (the message matches what I said in [this comment](https://github.com/scalacenter/bloop/pull/1190#issuecomment-601141571)). Sorry for the confusion! (I was confused myself 😅)
* [Pushed an additional commit](https://github.com/scalacenter/bloop/pull/1190/commits/ac876943e067a9f7383c7240975e8cf162610b38) to use a Java `ConcurrentHashMap` instead of a Scala `TrieMap`.

Let me know if I forgot something.

dotta commented:
Whohooo!! Thank YOU @jvican 🙌 

@olafurpg @tgodzik Would be fantastic if Metals could update its version of Bloop to include the changes in this PR (and, if possible, cut a release), as I think that's the easiest way to collect feedback from users. I'd for sure use it myself 😄

tgodzik commented:
@dotta If there is no changes to the way the launcher works, you can just change the Bloop version inside the settings to test it out.  We're working on bumping bloop version currently, we will follow up with the version with these changes later.

dotta commented:
@tgodzik Excellent. I've set the Bloop version in Metals to v1.4.0-RC1-142-90456e06 and compilation with Hydra is working fine. I have one question though, how can I correlate the SHA in the Bloop version (i.e., 90456e06) to a commit in Bloop? I'm asking because it's hard to know if the version of Bloop I'm using actually includes the changes I need. I'm just gussing it does because it was published on March 24th (I could find this very useful bit of information [here](https://scalacenter.github.io/bloop/docs/launcher-reference#install-the-launcher)). Thanks!
